src/hotspot/share/utilities/nativeStackPrinter.cpp

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+#include "runtime/frame.inline.hpp"
+#include "runtime/os.inline.hpp"
+#include "utilities/decoder.hpp"
+#include "utilities/globalDefinitions.hpp"
+#include "utilities/nativeStackPrinter.hpp"
+#include "utilities/ostream.hpp"
+
+bool NativeStackPrinter::print_stack(outputStream* st, char* buf, int buf_size,
+                                     address& lastpc, bool print_source_info,
+                                     int max_frames) {
+  if (os::platform_print_native_stack(st, _context, buf, buf_size, lastpc)) {
+    return true;
+  } else {
+    print_stack_from_frame(st, buf, buf_size, print_source_info, max_frames);
+    return false;
+  }
+}
+
+void NativeStackPrinter::print_stack_from_frame(outputStream* st, frame& fr,
```

Comment:
Why is `frame` passed by non-const reference?  This doesn't seem like a function that should be modifying
it?  Though changing this might also run afoul of const-correctness issues when accessing the frame, so
feel free to defer to a followup if that's a problem.

Reply:
I passed by reference here just to avoid adding in an additional level of copying in the new code as I was unsure if an additional copy might have an adverse impact. I never considered applying const and I also suspect it will have a significant flow-on effect if I try.

Reply:
Note that the contents in fr is overwritten in this function. All callers of this function must be careful because of that. I vote for fixing it before pushing this.

Reply:
@stefank - yes you are right. I think I introduced the reference in an earlier version of the code to avoid introducing an intermediary frame object, but as it stands the frame is passed in directly whereas the original code made a copy. I will fix it. Thanks.